packages/dataviews/src/types.ts

```diff
@@ -164,3 +202,78 @@ export interface ViewList extends ViewBase {
 }
 
 export type View = ViewList | ViewBase;
+
+export interface ActionBase< T extends Item > {
+	/**
+	 * Unique identifier of the action.
+	 *
+	 * For example, move-to-trash.
+	 */
+	id: string;
+
+	/**
+	 * User facing description of the action.
+	 *
+	 * For example, Move to Trash.
+	 */
+	label: string;
+
+	/**
+	 * Whether the action should be listed inline (primary)or in hidden in the more actions menu (secondary).
+	 */
+	isPrimary?: boolean;
+
+	/**
+	 * Icon to show for primary actions.
+	 *
+	 * It's required for a primary action, otherwise the action would be considered secondary.
+	 */
+	icon?: ReactNode;
+
+	/**
+	 * Whether the action can be performed for a given record.
+	 *
+	 * If not present, the action is considered to be eligible for all items. It takes the given
+	 * record as input.
+	 */
+	isEligible?: ( item: T ) => boolean;
+
+	/**
+	 * Whether the action can delete data, in which case the UI would communicate it via red color.
+	 */
+	isDestructive?: boolean;
+}
+
+export type WithRenderModal< T extends Item > = {
+	/**
+	 * The callback property is ignored when the `RenderModel` property is set.
+	 */
+	callback?: ( item: T ) => void;
+
+	/**
+	 * If an action requires that some UI be rendered in a modal.
+	 *
+	 * It can provide a component which takes as props the record as item and a closeModal function.
+	 *
+	 * When this prop is provided, the callback property is ignored.
+	 */
+	RenderModal?: ( props: { item: T; onClose: () => void } ) => ReactNode;
+};
+
+export type WithCallback< T extends Item > = {
+	/**
+	 * Callback function that takes the record as input and performs the required action.
+	 */
+	callback: ( item: T ) => void;
+};
+
+/**
+ * An action that can be performed on a record in the data set.
+ */
+export type Action< T extends Item = {} > = ActionBase< T > &
+	( WithRenderModal< T > | WithCallback< T > );
```

Comment:
This appropriately types the two possibilities of providing either a `RenderModal` or `callback`. It is allows the definition of `callback` when `RenderModal` is defined, though it has a note that it will not be used (as described in the documentation).